### ISSUE TITLE
CoffeeScript/Babel turning classes into functions on Cordova/Desktop

### ISSUE THREAD

**Issue by MatejJan** [meteor/meteor#13080]
When CoffeeScript is used in a project with a Cordova/Desktop (Electron) build target, all the classes are turned into functions for those builds. This creates problems when a CoffeeScript class tries to extend a JavaScript class.

This is not a problem on a normal web build (`Meteor.isModern` returning true) where CoffeeScript classes translate to JavaScript classes. However, in a Desktop build (`Meteor.isModern` returning false), the result of a class declaration is a function. 

Even worse, JavaScript classes (defined in .js files) do NOT get turned into functions in the Desktop build, leading to this incompatibility. However, if one includes a JavaScript class in a .cs file (using the embedded JavaScript syntax), that class will get turned into a function.

I'm assuming the problem then boils down to [these lines](https://github.com/meteor/meteor/blob/791f2e607bc83edc57a1186e58a6e746adb347ed/packages/non-core/coffeescript-compiler/coffeescript-compiler.js#L64C5-L69C72) in the CoffeeScript compiler:
```js
    // CoffeeScript contains a handful of features that output as ES2015+,
    // such as modules, generator functions, for…of, and tagged template
    // literals. Because they’re too varied to detect, pass all CoffeeScript
    // compiler output through the Babel compiler.
    const doubleRoastedCoffee =
      this.babelCompiler.processOneFileForTarget(inputFile, output.js);
```

Since Babel doesn't seem to be converting JavaScript classes in .js files into functions (even on Cordova/Desktop platforms), is there a way that the CoffeeScript compiler wouldn't be turning them either?

**Comment by nachocodoner** [comment]
I believe this is expected since when building a non-modern/legacy bundle, classes doesn't exist, and to proper functions are converted. But I can't see why uses legacy to compile in Cordova/Electron.

It would be nice to have a small reproduction repository where we can quickly try out to explore any ideas of possible solutions. Since I have more questions on the specific setup which I am not familiar with.

**Comment by MatejJan** [comment]
Here's a [reproduction repo](https://github.com/MatejJan/meteor-desktop-classes). 

Running with `meteor` and loading in the browser should give the default project with two extra lines added:

> The parent JS class returns parent.
> The child CS class returns child + parent.

Now, in addition, also start `npm run desktop`. In the desktop app that should open, the child line will not be correct. Open the console and you will see:
```
Exception in template helper: TypeError: Class constructor Parent cannot be invoked without 'new'
    at new Child (meteor://desktop/__cordova/packages/test.js?hash=bff098a…:83:25)
    at Object.child (meteor://desktop/__cordova/app/app.js?hash=3b482fc…:116:19)
    at meteor://desktop/__cordova/packages/blaze.js?hash=dbc770c…:2849:16
    at meteor://desktop/__cordova/packages/blaze.js?hash=dbc770c…:1512:16
    at meteor://desktop/__cordova/packages/blaze.js?hash=dbc770c…:2894:66
    at Function.Template._withTemplateInstanceFunc (meteor://desktop/__cordova/packages/blaze.js?hash=dbc770c…:3567:14)
    at meteor://desktop/__cordova/packages/blaze.js?hash=dbc770c…:2893:27
    at Spacebars.call (meteor://desktop/__cordova/packages/spacebars.js?hash=8047676…:171:18)
    at Spacebars.mustacheImpl (meteor://desktop/__cordova/packages/spacebars.js?hash=8047676…:105:25)
    at Object.Spacebars.mustache (meteor://desktop/__cordova/packages/spacebars.js?hash=8047676…:109:39)
```

This is because the source code (parent.js and child.coffee):
```js
class Parent {
    test() {
        return 'parent';
    }
}

Test.Parent = Parent;
```
```coffee
class Test.Child extends Test.Parent
    test: ->
        parent = super arguments...
        "child + #{parent}"
```

got turned into:
```js
(function(){

///////////////////////////////////////////////////////////////////////
//                                                                   //
// packages/test/parent.js                                           //
//                                                                   //
///////////////////////////////////////////////////////////////////////
                                                                     //
class Parent {
    test() {
        return 'parent';
    }
}

Test.Parent = Parent;

///////////////////////////////////////////////////////////////////////

}).call(this);






(function(){

///////////////////////////////////////////////////////////////////////
//                                                                   //
// packages/test/child.coffee                                        //
//                                                                   //
///////////////////////////////////////////////////////////////////////
                                                                     //
__coffeescriptShare = typeof __coffeescriptShare === 'object' ? __coffeescriptShare : {}; var share = __coffeescriptShare;
function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; _setPrototypeOf(subClass, superClass); }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function () { function _setPrototypeOf(o, p) { o.__proto__ = p; return o; } return _setPrototypeOf; }(); return _setPrototypeOf(o, p); }
Test.Child = /*#__PURE__*/function (_Test$Parent) {
  _inheritsLoose(Child, _Test$Parent);
  function Child() {
    return _Test$Parent.apply(this, arguments) || this;
  }
  var _proto = Child.prototype;
  _proto.test = function () {
    function test() {
      var parent;
      parent = _Test$Parent.prototype.test.apply(this, arguments);
      return "child + " + parent;
    }
    return test;
  }();
  return Child;
}(Test.Parent);
///////////////////////////////////////////////////////////////////////

}).call(this);
```

Meanwhile in the browser everything is OK:
```js
(function(){

///////////////////////////////////////////////////////////////////////
//                                                                   //
// packages/test/parent.js                                           //
//                                                                   //
///////////////////////////////////////////////////////////////////////
                                                                     //
class Parent {
    test() {
        return 'parent';
    }
}

Test.Parent = Parent;

///////////////////////////////////////////////////////////////////////

}).call(this);






(function(){

///////////////////////////////////////////////////////////////////////
//                                                                   //
// packages/test/child.coffee                                        //
//                                                                   //
///////////////////////////////////////////////////////////////////////
                                                                     //
__coffeescriptShare = typeof __coffeescriptShare === 'object' ? __coffeescriptShare : {}; var share = __coffeescriptShare;
Test.Child = class Child extends Test.Parent {
  test() {
    var parent;
    parent = super.test(...arguments);
    return "child + ".concat(parent);
  }
};
///////////////////////////////////////////////////////////////////////

}).call(this);
```

One thing I can add to the first message is that the reason javascript files weren't transpiled into functions was because I didn't include the ecmascript package in the test package. If I do, that one will get transpiled as well. But that isn't a solution, since the problem still exists when trying to use an NPM package that use ES6 classes.

To sum up, the ideal solution here is that CoffeeScript classes remain compiled as javascript classes and not get turned into functions in the Cordova/Desktop build.

**Comment by MatejJan** [comment]
As a [temporary workaround](https://github.com/Retronator/meteor/commit/d61dba511fe810d07bb594b6cf0234a232e963d5), forcing `modernBrowsers` in the babel compiler (via a custom fork of Meteor) works fine and produces the same codebase for browsers and Cordova/Desktop.
